### PR TITLE
[dynamo] Reorder logs

### PR DESCRIPTION
**Pull request by angelayi**
Currently when there is a print/warning in the graph, dynamo graph breaks causing export to fail. However export would like to just skip over these print/warning calls: https://github.com/pytorch/pytorch/issues/113792.

Additionally there's a torch.compile feature request to "reorder prints" so that instead of graph breaking when hitting prints/logging, we can skip over these prints to create larger compiled graphs, and then print the results out after those compiled graphs: https://github.com/pytorch/pytorch/issues/93739. This PR also adds the `reorderable_logging_functions` config for users to register logging functions to be reordered (like `print` or a custom logging function). Printout of the bytecode after reordering the prints looks like the following: P914736600

There are some limitations to the printing right now:
* You can only register logging functions, not methods
* Inputs to the logging functions can only be tensors, constants, and format strings
* Inputs to the logging functions which will later be mutated in-place will not be printed correctly

TODO: Add the following tests
* print function with argument of nested data structure;
* print function with argument of nested data structure being updated inside of compile region (this would test if we handle side effect correctly);
* custom defined logging functions with nn.Module or nn.Module attribute arguments;
* custom defined logging functions with submodule input/output as arguments (we need to handle the mapping and fused-out value);
* custom defined logging functions with tensor argument and mutation inside of the function (TBD: this may increase memory usage);

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_misc.py [hunk 1]
```diff
@@ -7487,6 +7488,71 @@ def forward(self, input):
...
+    @torch._dynamo.config.patch(reorder_prints=True)
+    def test_reorder_print_graph_break(self):
+        def f(x):
+            x1 = x + x
+            print(f"res: {x1}")
+            x2 = x1 * x1
+            torch._dynamo.graph_break()
+            x3 = x2 + x2
+            print(1, 2, 3)
+            return (x1, x3)
```
Comment (williamwen42) [inline test/dynamo/test_misc.py:7544]: Why do we also return `x1`? I think we should test only returning `x3` and expect that the resulting code should still work (and I confirm that it does).

**Review by williamwen42 — COMMENTED** [review]
A few things we should also have:
- support reordering `logging` calls - we could do this in a separate PR (you mentioned that you ran into skipfiles issues)
- support print statements involving variables that will be mutated, e.g.
```python
import torch
def f(x):
    print("start")
    x1 = x + x
    x1.add_(1)
    print(x1)
    x1.add_(1)
    x2 = x1 * x1
    print(1, 2, 3)
    x3 = x2 + x2
    return x3

x = torch.ones(3, 3)
opt_f = torch.compile(backend="eager")(f)
opt_f(x)
```
We could do this by copying/cloning the args.

**Comment by jansel** [comment]
Tests are failing

File: test/dynamo/test_misc.py [hunk 2]
```diff
@@ -7487,6 +7488,71 @@ def forward(self, input):
...
+            print("start")
+            x1 = x + x
+            print(x1)
+            x2 = x1 * x1
+            print(1, 2, 3)
+            x3 = x2 + x2
+            return (x1, x3)
+
+        x = torch.ones(3, 3)
+        opt_f = torch.compile(backend="eager")(f)
```
Comment (jansel) [inline test/dynamo/test_misc.py:7503]:
```suggestion
        opt_f = torch.compile(backend="eager", fullgraph=True)(f)
```

File: test/dynamo/test_misc.py [hunk 3]
```diff
@@ -7487,6 +7488,71 @@ def forward(self, input):
...
+            x1 = x + x
+            warnings.warn("start")
+            x2 = x1 * x1
+            warnings.warn(f"{x2}")
+            x3 = x2 + x2
+            return x3
+
+        x = torch.ones(3, 3)
+        torch.export.export(f, (x,))
+        opt_f = torch.compile(backend="eager")(f)
```
Comment (jansel) [inline test/dynamo/test_misc.py:7526]:
```suggestion
        opt_f = torch.compile(backend="eager", fullgraph=True)(f)
```

File: torch/_dynamo/output_graph.py [hunk 4]
```diff
@@ -929,6 +929,14 @@ def append_prefix_insts():
                 graph_output_var,
                 tempvars={val: None for val, count in pass1.uses.items() if count > 1},
             )
+
+            if torch._dynamo.config.reorder_prints and len(tx.debug_locals) != 0:
+                for debug_var, args in tx.debug_locals:
+                    pass2(debug_var)
```
Comment (jansel) [inline torch/_dynamo/output_graph.py:935]: I think you may need to do this stuff in pass1 as well if any temporary variables need to get constructed.

File: torch/_dynamo/variables/user_defined.py [hunk 5]
```diff
@@ -854,6 +854,27 @@ def odict_getitem(self, tx, key):
...
+
+    def call_function(self, tx, args, kwargs):
+        if tx.export:
+            # For export cases, we can just make debugging functions no-ops
+            return
+        elif torch._dynamo.config.reorder_prints:
+            tx.debug_locals.append((self, list(args)))
+            return
+        else:
+            return super().call_function(tx, args, kwargs)
```
Comment (jansel) [inline torch/_dynamo/variables/user_defined.py:872]: I think this will just call `unimplemented()`.  Let's change the error message to suggest people cal set `config.reorder_prints=True` to fix it.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/output_graph.py [hunk 6]
```diff
@@ -934,6 +934,13 @@ def append_prefix_insts():
         else:
             graph_output_var = self.new_var("graph_out")
             pass1 = PyCodegen(tx, root, graph_output_var)
+            if torch._dynamo.config.reorder_prints and len(tx.debug_locals) != 0:
+                for debug_var, args in tx.debug_locals:
+                    pass1(debug_var)
+                    for arg in args:
+                        pass1(arg)
+                    pass1.extend_output(create_call_function(len(args), True))
```
Comment (yanboliang) [inline torch/_dynamo/output_graph.py:942]: Let's refactor out a helper function?

File: torch/_dynamo/variables/user_defined.py [hunk 7]
```diff
@@ -873,6 +873,31 @@ def odict_getitem(self, tx, key):
         )(collections.OrderedDict.__getitem__(self.value, key.as_python_constant()))
 
 
+class DebuggingVariable(UserDefinedObjectVariable):
```
Comment (yanboliang) [inline torch/_dynamo/variables/user_defined.py:876]: I think this wraps a function, so probably we should not extend from UDO. This variable can be python constant, but UDO can't.

File: torch/_dynamo/variables/user_defined.py [hunk 8]
```diff
@@ -873,6 +873,31 @@ def odict_getitem(self, tx, key):
...
+            return
+        elif torch._dynamo.config.reorder_prints:
+            tx.debug_locals.append((self, list(args)))
+            return
+        else:
+            unimplemented(
+                f"Debugging functions like {self} are not supported by Dynamo. "
+                "Please set `config.reorder_prints=True` to fix it."
+            )
+            return super().call_function(tx, args, kwargs)
```
Comment (yanboliang) [inline torch/_dynamo/variables/user_defined.py:895]: If graph break, L895 is not needed.

File: torch/export/_unflatten.py [hunk 9]
```diff
@@ -0,0 +1,883 @@
+import abc
```
Comment (yanboliang) [inline torch/export/_unflatten.py:1]: Probably we can split the changes into two separate PRs: one for dynamo and another one for export.

**Review by yanboliang — COMMENTED** [review]
Thanks for working on this feature!

File: torch/export/_unflatten.py (hunk 9 above)
Comment (angelayi) [reply torch/export/_unflatten.py:1]: oops i think i accidentally committed this

File: torch/_dynamo/config.py [hunk 10]
```diff
@@ -370,6 +372,27 @@ def default_debug_dir_root():
...
+    logging.debug,
+    logging.error,
+    logging.exception,
+    logging.info,
+    logging.log,
+    logging.warning,
+    print,
+    warnings.warn,
+}
+
```
Comment (yanboliang) [inline torch/_dynamo/config.py:395]: I think probably one flag is better then two, should we only keep the ```reorderable_logging_functions``` and None by default? Users can set it if they want to reorder any function rather than graph break on it.

File: torch/_dynamo/output_graph.py [hunk 11]
```diff
@@ -999,6 +994,19 @@ def cleanup_graph(self):
...
+                codegen(debug_var)
+                for arg in args:
+                    codegen(arg)
+                codegen.extend_output(create_call_function(len(args), True))
+
+        self.side_effects.codegen_hooks(codegen)
+        self.side_effects.codegen_save_tempvars(codegen)
+        codegen.restore_stack(stack_values, value_from_source=not tx.export)
+        self.side_effects.codegen_update_mutated(codegen)
+
```
Comment (yanboliang) [inline torch/_dynamo/output_graph.py:1009]: I think we should move L998-L1003 down to after L1008, probably there are some variables refer to temp vars or mutated vars, though they should be basic types, but they can be arguments of these log functions. You can add unit tests for such cases as follow-ups.

**Comment by yanboliang** [comment]
Looks pretty close, thanks for working on this!

File: torch/_dynamo/output_graph.py (hunk 11 above)
Comment (angelayi) [reply torch/_dynamo/output_graph.py:1009]: I can't add it all the way to the end, because I will add outputs to the codegen. So to my understanding this needs to go before the `codegen.restore_stack` call.

**Review by yanboliang — COMMENTED** [review]
LGTM other than two minor issues!

File: test/dynamo/test_autograd_function.py [hunk 12]
```diff
@@ -271,7 +271,7 @@ def test_print_in_bwd(self):
         opt_model = torch._dynamo.optimize("eager", nopython=True)(model)
         x = torch.randn(2, 2, dtype=torch.double, requires_grad=True)
         with self.assertRaisesRegex(
-            torch._dynamo.exc.Unsupported, ".*BuiltinVariable\\(print\\).*"
+            torch._dynamo.exc.Unsupported, r".*Debugging functions.*"
```
[the remark this answers is not in the document]
Comment (angelayi) [reply test/dynamo/test_autograd_function.py:274]: yeah I think some previous changes caused me to have to change this place 😅 but not anymore

**Review by yanboliang — APPROVED** [review]

File: torch/_dynamo/output_graph.py [hunk 13]
```diff
@@ -999,6 +994,19 @@ def cleanup_graph(self):
...
 
+    def construct_python_bytecode(self, tx, codegen: PyCodegen, stack_values):
+        self.side_effects.codegen_hooks(codegen)
+        self.side_effects.codegen_save_tempvars(codegen)
+
+        for debug_var, args in tx.debug_locals:
+            codegen(debug_var)
+            for arg in args:
+                codegen(arg)
+            codegen.extend_output(create_call_function(len(args), True))
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:1005]: What exactly is this doing?

File: torch/_dynamo/variables/builder.py [hunk 14]
```diff
@@ -471,6 +472,11 @@ def build_key_value(i, k, v):
         elif isinstance(value, enum.Enum):
             self.install_guards(GuardBuilder.ID_MATCH)
             return EnumVariable(value=value, source=self.source)
+        elif DebuggingVariable.is_reorderable_logging_function(value):
+            # Put this above builtin_callable so that print() can be handled
+            # along with other builtin debugging functions
+            self.install_guards(GuardBuilder.BUILTIN_MATCH)
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:478]: You've made the list of reorderable logging functions customizable, why is BUILTIN_MATCH the correct guard to use here then?

File: torch/_dynamo/output_graph.py (hunk 13 above)
Comment (yanboliang) [reply torch/_dynamo/output_graph.py:1005]: This is generating the call logging function bytecode at the end of the graph.

File: torch/_dynamo/variables/builder.py (hunk 14 above)
Comment (yanboliang) [reply torch/_dynamo/variables/builder.py:478]: This can be ```FUNCTION_MATCH``` as well, but they are equivalent IIFC.

**Review by angelayi — COMMENTED** [review]
lint..

**Comment by angelayi** [comment]
@pytorchot merge
